### PR TITLE
`Lectures`: Remove redundant competency management from guided creation mode

### LINKED ISSUE #8862: Competencies incorrectly linking to lecture units in ecture creaion guided mode.
### Describe the bug

When creating a new Lecture Unit during Lecture creation in guided mode. 
There exists the possibility to link competencies to the Lecture Unit in creation.
After finishing the creation of the Lecture unit, the selected linked competencies are not remembered.

Duplicat functionality is found in the next step, however lining competencies here overwrites existing links.

### To Reproduce

1. Create a competency
2. Enter Lecture creation
3. Switch to Guided Mode
4. Continue to step 4 of Lecture creation
5. Create any new Lecture Unit
6. During Lecture Unit Creation select competency to link
7. Finish lecture creation.
8. Go to the competency and find Lecture Unit not added

1. Create a competency
2. Link at least 1 Lecture Unit
2. Enter Lecture creation
3. Switch to Guided Mode
4. Continue to step 4 of Lecture creation
5. Create any new Lecture Unit
6. Continue to next step
7. Link Competency to the newly created Lecture Unit
9. Finish lecture creation
10. Go to the competency and find prevously linked units missing

### Expected behavior

The Lecture Unit beeing added or old linked units not beeing removed.

### Screenshots

_No response_

### Which version of Artemis are you seeing the problem on?

7.3.0

### What browsers are you seeing the problem on?

Firefox

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by N0W0RK**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

During the lecture creation guided mode, a lecture and its units and attachemnts are created step by step. 
The linking of compentencies however is duplicated. 
It can be either done during lecture unit creation by setting a checkmark on the appropiate competency. 
It is also possible to configure the linking competencies in the next step of guided mode. 
Here it is possible to create edit and delete competencies. 

This is redundant behaviour inside the lecture creation mode. 
Further more the creation and deletion of competencies is an action not really needed during lecture creation.

Fixes #8862 

### Description
This PR removes the competency step from the lecture wizard. 
Additionally the linking of competencies to units during the wizard is fixed. 
Previously competencies weren't linked to the lecture units even if selected in the wizard.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to lectures
4. Select `Create Lecture`
5. Select `Switch to guided mode`
6. Enter a name and continue to the `Ùnits` step
7. Create a unit and link competencies
8. Ensure that this is the last step. Clicking on `Finish` should fully create the lecture
9. Unsure that all created units are properly linked to competencies

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| lecture-update-wizard.component.ts | 100% | ✅ |
| lecture-wizard-competencies.component.ts | not found (deleted) | ❌ |
| lecture-wizard-units.component.ts | 100% | ✅ |

### Screenshots

**Guided Lecture Creation without seperate competancy management:**
image

**Updated Progressbar**
image

## Summary by CodeRabbit

- **New Features**
  - Updated step visibility in the lecture update wizard for a more dynamic user experience.
  - Added competency tracking to various unit types (Text, Video, Online, Attachment) within the lecture update wizard.

- **Bug Fixes**
  - Removed the unused competencies step from the wizard component to streamline the flow.

- **Tests**
  - Enhanced test coverage to include competency-related functionality for unit types.

### CODE REVIEW

File: src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.ts [hunk 1]
```diff
@@ -21,7 +20,10 @@ export class LectureUpdateWizardComponent implements OnInit {
     @Input() isSaving: boolean;
 
     @ViewChild(LectureUpdateWizardUnitsComponent, { static: false }) unitsComponent: LectureUpdateWizardUnitsComponent;
-    @ViewChild(LectureUpdateWizardCompetenciesComponent, { static: false }) competenciesComponent: LectureUpdateWizardCompetenciesComponent;
+
+    readonly LECTURE_UPDATE_WIZARD_TITLE_STEP = 1;
+    readonly LECTURE_UPDATE_WIZARD_PERIOD_STEP = 2;
+    readonly LECTURE_UPDATE_WIZARD_UNIT_STEP = 4;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.ts:26]: Can you also use the variables in the html file?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.ts [hunk 2]
```diff
@@ -62,7 +71,7 @@ export class LectureUpdateWizardComponent implements OnInit {
      * Progress to the next step of the wizard mode
      */
     next() {
-        if (this.currentStep === 2 || this.currentStep === 5) {
+        if (this.currentStep === 2 || this.currentStep === 4) {
```
Comment (Strohgelaender) [inline src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.ts:74]: magic numbers

File: src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.html [hunk 3]
```diff
@@ -24,38 +21,32 @@
...
-                    [isPerformed]="isCompleted(2)"
+                    [currentlySelected]="isCurrent(LECTURE_UPDATE_WIZARD_PERIOD_STEP)"
+                    [isPerformed]="isCompleted(LECTURE_UPDATE_WIZARD_PERIOD_STEP)"
                     [labelTranslationKey]="'artemisApp.lecture.wizardMode.steps.periodStepTitle'"
                 />
                 <div class="connector" [ngClass]="isCompleted(2) ? 'check' : 'unset'"></div>
                 <jhi-lecture-update-wizard-step
-                    [currentlySelected]="isCurrent(3)"
-                    [isPerformed]="isCompleted(3)"
+                    [currentlySelected]="isCurrent(LECTURE_UPDATE_WIZARD_ATTACHMENTS_STEP)"
```
Comment (Strohgelaender) [inline src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.html:36]: Does the `isCurrent` method just check the `currentStep`  Attribute? If yes, please replace this function call by a direct conditional. This enables the angular change detection to smartly evaluate the condition only on changes for `currentStep` whereas methods must be called every time. (See client guides for more)

Comment (N0W0RK) [reply src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.html]: I addressed this.

However I could not find any reference to this in the client guidelines.

**Review by Strohgelaender — COMMENTED** [review]
Thanks for including my suggestions and improving the code. It now looks good to me, but the client tests are failing due to coverage (probally about the lines 53 to 59 in the update wizard?). Please add some tests here.

File: src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.html (hunk 3 above)
Comment (Strohgelaender) [reply src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.html]: It's the warning in the `0. General` section. (I also experienced that for some reason these big warnings actually get overlooked more easily than the normal text)

**Comment by N0W0RK** [comment]
All lecture wizard components have 100% line coverage except the `lecture-update-wizard-step.component.ts`. That has only 60% however I am honestly not sure how to increase the coverage there.

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by Strohgelaender — APPROVED** [review]
code lgtm

**Review by florian-glombik — APPROVED** [review]
Tested on ts2, the described functionality works as expected - configuring the `Units` is now the last step instead of `Competencies`

While reviewing, I found a couple of issues that are not in the scope of the PR:
- [Lectures: Channel name is cleared after completing second step in guided mode #9167](https://github.com/ls1intum/Artemis/issues/9167)
- [Lectures: Attachment upload in guided mode does not work #9168](https://github.com/ls1intum/Artemis/pull/8989)
- [Lectures: Save input from guided modes only when last step is completed #9169](https://github.com/ls1intum/Artemis/issues/9169)

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by Strohgelaender — APPROVED** [review]
re-approve code

**Review by JohannesWt — APPROVED** [review]
Reapprove

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge

**Review by JohannesWt — DISMISSED** [review]
Reapprove

File: src/test/javascript/spec/component/import/exercise-import.component.spec.ts [hunk 4]
```diff
@@ -11,6 +11,7 @@ import { CodeAnalysisPagingService } from 'app/exercises/programming/manage/serv
 import { ProgrammingExercisePagingService } from 'app/exercises/programming/manage/services/programming-exercise-paging.service';
 import { QuizExercisePagingService } from 'app/exercises/quiz/manage/quiz-exercise-paging.service';
 import { ExerciseImportComponent } from 'app/exercises/shared/import/exercise-import.component';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (MaximilianAnzinger) [inline src/test/javascript/spec/component/import/exercise-import.component.spec.ts:14]: Why do we need this?

Comment (N0W0RK) [reply src/test/javascript/spec/component/import/exercise-import.component.spec.ts:14]: I wish I knew.
Removing it, causes eslint to fail claiming `PagingService` is not used as a type. 
This however contradicts [line 265](https://github.com/ls1intum/Artemis/blob/8349c46a8f26b7f3594282aec62bd6580fa4275b/src/test/javascript/spec/component/import/exercise-import.component.spec.ts#L265) and removing the import causes errors.

Comment (MaximilianAnzinger) [reply src/test/javascript/spec/component/import/exercise-import.component.spec.ts:14]: Removing the line works fine for me.

**Review by JohannesWt — DISMISSED** [review]
Reapprove once more :)

**Review by florian-glombik — DISMISSED** [review]
Re-tested on ts3

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove

**Review by JohannesWt — DISMISSED** [review]
Code

**Review by JohannesWt — DISMISSED** [review]
Reapprove after merge conflict

**Review by florian-glombik — DISMISSED** [review]
Re-approve after merge conflict

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove after merge conflict

**Review by Strohgelaender — APPROVED** [review]

**Review by JohannesWt — APPROVED** [review]
Code reapprove - let's do it this time 👍

**Review by florian-glombik — APPROVED** [review]
Re-approve after merge conflict

**Review by JohannesStoehr — APPROVED** [review]
Reapprove

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**